components/datetime/src/any.rs

```diff
@@ -0,0 +1,250 @@
+/// heavy construction of [`DateTimeFormat`], and then fast formatting of [`DateTime`] data using the instance.
+///
+/// [`icu_datetime`]: crate
+/// [`DateTimeFormat`]: crate::datetime::DateTimeFormat
+///
+/// # Examples
+///
+/// ```
+/// use icu::calendar::{any_calendar::AnyCalendar, DateTime, Gregorian};
+/// use icu::datetime::{options::length, any::AnyDateTimeFormat};
+/// use icu::locid::Locale;
+/// use icu_provider::any::DynProviderAnyMarkerWrap;
+/// use std::str::FromStr;
+///
+/// let provider = icu_testdata::get_provider();
+///
+/// let mut options = length::Bag::from_date_time_style(length::Date::Medium, length::Time::Short);
+///
+/// let dtf = AnyDateTimeFormat::try_new_with_buffer_provider(Locale::from_str("en-u-ca-gregory").unwrap(), &provider, &options.into())
+///     .expect("Failed to create DateTimeFormat instance.");
+///
+/// let datetime = DateTime::new_gregorian_datetime_from_integers(2020, 9, 1, 12, 34, 28, 0)
+///     .expect("Failed to construct DateTime.");
+/// let any_datetime = datetime.to_any();
+///
+/// let value = dtf.format_to_string(&any_datetime).expect("calendars should match");
+/// assert_eq!(value, "Sep 1, 2020, 12:34 PM");
+/// ```
+///
+/// This model replicates that of `ICU` and `ECMA402`.
+///
+/// [data provider]: icu_provider
+pub struct AnyDateTimeFormat(pub(super) raw::DateTimeFormat, AnyCalendar);
+
+impl AnyDateTimeFormat {
+    /// Construct a new [`AnyDateTimeFormat`] from a data provider that implements
+    /// [`AnyProvider`].
+    ///
+    /// The provider must be able to provide data for the following keys: `datetime/symbols@1`, `datetime/lengths@1`,
+    /// `datetime/symbols@1`, `datetime/skeletons@1`, `datetime/week_data@1`, and `plurals/ordinals@1`.
+    ///
+    /// Furthermore, based on the type of calendar used, one of the following data keys may be necessary:
+    ///
+    /// - `u-ca-japanese` (Japanese calendar): `calendar/japanese@1`
+    #[inline]
+    pub fn try_new_with_any_provider<T: Into<Locale>, P>(
+        locale: T,
+        data_provider: &P,
+        options: &DateTimeFormatOptions,
+    ) -> Result<Self, DateTimeFormatError>
+    where
+        P: AnyProvider,
+    {
+        let locale = locale.into();
+
+        let kind = AnyCalendarKind::from_locale(&locale)
+            .ok_or(DateTimeFormatError::MissingCalendarOnLocale)?;
+
+        let calendar = AnyCalendar::try_new_with_any_provider(kind, data_provider)?;
+
+        let downcasting = data_provider.as_downcasting();
+        Ok(Self(
+            raw::DateTimeFormat::try_new(locale, &downcasting, options)?,
+            calendar,
+        ))
+    }
+
+    /// Construct a new [`AnyDateTimeFormat`] from a data provider that implements
+    /// [`BufferProvider`].
+    ///
+    /// The provider must be able to provide data for the following keys: `datetime/symbols@1`, `datetime/lengths@1`,
+    /// `datetime/symbols@1`, `datetime/skeletons@1`, `datetime/week_data@1`, and `plurals/ordinals@1`.
```

Comment:
Nit, here and above and below: These docs are not good because they will quickly get stale. Better is to add the unstable constructor, and say that the provider must be able to provide data for the keys listed in the signature of the unstable constructor.

Reply:
It might be worth listing both and adding a comment on the unstable bounds asking people to keep stuff in sync?

We could just say "all the types here" but we also need to be able to talk about the mandatory keys vs the per-calendar keys. I'm not 100% sure what I prefer here.

Reply:
Also, these keys are generally considered internal. Most users won't know what to do when they read "make sure these keys are available". They'll either be using the full data file or one of our automated tools to slice it for them.